### DISCUSSION TITLE
The road towards the first upgrade (and beyond)

### DISCUSSION THREAD

**Discussion by kostko** [oasisprotocol/community-forum#5] [Mainnet]
## The Road Towards the First Upgrade...

While Mainnet was successfully launched three weeks ago, this is only the beginning. Thinking ahead we have identified several areas that need improvements in order to make [ParaTimes and the overall network](https://docs.oasis.dev/general/oasis-network/overview) even more capable. In the near-term the following are a set of features the Foundation proposes to deploy as a Mainnet upgrade in Q1 2021. Each of those links to technical details in the form of an [ADR](https://github.com/oasisprotocol/oasis-core/blob/master/docs/adr/index.md). Many of the proposed changes have already been implemented in [Oasis Core](https://github.com/oasisprotocol/oasis-core) and some are undergoing audits.

#### Light Clients and Checkpoint Sync

In order to make bootstrapping of new network nodes much faster, the upgrade will introduce support for light clients and restoring state from checkpoints provided by other nodes in the network (see [oasis-core#2880](https://github.com/oasisprotocol/oasis-core/issues/2880) and [oasis-core#2440](https://github.com/oasisprotocol/oasis-core/issues/2440)). Nodes will be able to announce that they provide public light client endpoints to make discovery easier (e.g., enabling block explorers to publish such endpoints).

#### Random Beacon

The random beacon is used by the consensus layer for ParaTime committee elections and is a critical component in providing security for ParaTimes with an open admission policy. [ADR 0007](https://github.com/oasisprotocol/oasis-core/blob/master/docs/adr/0007-improved-random-beacon.md) specifies a random beacon implementation based on [SCRAPE](https://eprint.iacr.org/2017/216.pdf) which provides unbiased output as long as at least one participant (validator node) is honest.

#### On-chain Governance for Easier Upgrade Coordination

So far all of the network upgrades had to be manually coordinated off-chain, validators needed to take dumps at specific heights, patch the dump etc. Every upgrade also required any previous state (and history) to be wiped. The new on-chain governance service as specified by [ADR 0006](https://github.com/oasisprotocol/oasis-core/blob/master/docs/adr/0006-consensus-governance.md) provides a simple framework for submitting governance proposals, validators voting on proposals and once an upgrade proposal passes, having a way to perform the upgrade in a controlled manner which minimizes downtime.

#### ROSE Transfers Between the Consensus Layer and ParaTimes

In the current Mainnet there is no way for ParaTimes to interact with other accounts in the consensus layer. [ADR 0003](https://github.com/oasisprotocol/oasis-core/blob/master/docs/adr/0003-consensus-runtime-token-transfer.md) proposes a mechanism where ParaTimes can emit messages as part of processing any ParaTime block. These messages can trigger operations in the consensus layer on the ParaTime's behalf. This also means that ParaTimes get their own accounts in the consensus layer which can hold and transfer tokens.

#### A Path Towards Self-governing ParaTimes

Currently all ParaTimes can only be governed by a single entity -- the ParaTime owner. In this regard governance means being able to update certain fields in the ParaTime descriptor stored by the consensus layer registry service. On one hand the ParaTime descriptor contains security-critical parameters and on the other there needs to be a mechanism through which the ParaTimes can be upgraded (especially so for TEE-based runtimes where a specific runtime binary is enforced via remote attestation mechanisms). [ADR 0004](https://github.com/oasisprotocol/oasis-core/blob/master/docs/adr/0004-runtime-governance.md) extends ParaTime governance options and enables a path towards ParaTimes that can define their own governance mechanisms.

## ...and Beyond

Beyond the consensus layer updates there are also other areas that the Foundation is thinking about based on feedback from the community which are in early phases:

* Improving the ParaTime developer experience by introducing a high-level ParaTime SDK that provides common functionalities.
* Improving the frontend developer experience by introducing a JavaScript SDK that supports both the consensus layer and arbitrary ParaTimes based on the ParaTime SDK.
* Building a bridge between ParaTimes and other networks like Ethereum.

We welcome any additional proposals for enhancements from the community (either via the [contribution process in Oasis Core](https://github.com/oasisprotocol/oasis-core/blob/master/CONTRIBUTING.md) or via high-level suggestions in [this community forum](https://github.com/oasisprotocol/community)) and are also [providing grants](https://oasisprotocol.org/ecosystem-grants).

**Comment by Afauvre** [comment]
I would also propose increasing the validator set from 80 to 100. 

There are a number of community-driven validators that are just below the the 80 node validator set now, they've been active for some time and I believe it will be critical to expanding the growing community around the Oasis Network.

**Comment by keefertaylor** [comment]
Tessellated Geometry supports expanding the validator set. 

We've seen other Tendermint systems easily scale to 100 validators and if we have good high quality validators who can help decentralize the system further and expand the community we should absolutely include the.

**Comment by joesixpack** [comment]
100 is still very, very small in terms of decentralization and given that Tendermint is BFT/DPoS with a low 33% vulnerability, the more validators, the merrier.

**Comment by Mr-Cansky** [comment]
1. Add more usecase for holder
2. Focus on Dex ecosystem & Bridge
3. Make a first project built on Rose

**Comment by tjanez** [comment]
Let's move further discussion to #24.
